### PR TITLE
Add buffer to over and hybrid methods

### PR DESCRIPTION
**Pull request by josephenguehard**
As noted in the README, `OverSampler` and `HybridSampler` yield repeated samples one after the other. 

This PR introduces a `buffer` to solve this limitation:

Defining:
```python
import collections
import torch


class Dataset(torch.utils.data.IterableDataset):
    def __init__(self):
        super(Dataset, self).__init__()
        self.data = list(range(20))
        self.labels = [1] * 12 + [0] * 8
        
    def __iter__(self):
        for x, y in zip(self.data, self.labels):
            yield x, y

dataset = Dataset()
```

`OverSampler` without buffer:
```python
sample = OverSampler(
    dataset=dataset,
    desired_dist={0: .5, 1: .5},
    seed=42,
)

y_dist = collections.Counter()

batches = torch.utils.data.DataLoader(sample, batch_size=1)
outputs = list()
for xb, yb in batches:
    outputs.append(xb.item())

>>> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 12, 12, 12, 12, 12, 12, 12, 12, 13, 13, 13, 14, 14, 14, 15, 15, 15, 16, 16, 18, 18, 18, 18, 19, 19]
```

`OverSampler` with buffer:
```python
sample = OverSampler(
    dataset=dataset,
    desired_dist={0: .5, 1: .5},
    buffer_size=10,
    seed=42,
)

y_dist = collections.Counter()

batches = torch.utils.data.DataLoader(sample, batch_size=1)
outputs = list()
for xb, yb in batches:
    outputs.append(xb.item())

>>> [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 12, 12, 12, 12, 12, 12, 12, 12, 13, 12, 13, 14, 13, 12, 12, 13, 14, 14, 13, 14, 12, 16, 13, 15, 17, 14, 18, 12]
```

`HybridSampler` can be compared the same way.

### CODE REVIEW

**Review by MaxHalford — CHANGES_REQUESTED** [review]
Hey there, thanks for looking into this!

I wouldn't do things this way. The issue is that samples that are repeated are not yielded in a random order. You propose to handle the randomization internally.

I think that the randomization should be done externally. There could be a `ShuffleBuffer` class which inherits from `IterableDataset`. It takes as input and wraps `IterableDataset`. It handles the randomization. Therefore, we can plug any of these samplers into it. 

Sorry for being brief. Does it make sense to you?
